### PR TITLE
WIP: Don't draw_idle in non-GUI backends

### PR DESCRIPTION
**Pull request by mdboom**
Consider this a "proposed" PR, since I have a feeling the solution can't possibly be this easy.

While doing some timings on image drawing, I just discovered that running something like the following:

```python
import matplotlib
matplotlib.use('Agg')
from matplotlib import pyplot as plt

plt.plot([1,2,3])
plt.savefig("test.png")
```

results in the entire figure being drawn twice.  Once to save the file, and then again as the result of `draw_idle`.  It seems that `draw_idle` really shouldn't do anything in a non-GUI backend.

Making `draw_idle` a no-op in the canvas base class seems to address this, and all the GUI backends override `draw_idle` anyway.  This behavior seems to have been here for a very long time.

I can't seem to see how this breaks things, by @tacaswell has been down in the weeds of `draw_idle` more recently...  Maybe he has some ideas as to why this is necessary.

### CODE REVIEW

**Comment by tacaswell** [comment]
.... well fine that does make more sense that the epicycles I added to make it only draw _one_ extra time.

There is a bunch more logic that should be ripped out and I suspect the failing tests just need to have explicit `draw` calls added to force updates.

**Comment by tacaswell** [comment]
just to be clear, the tone in my last comment should be mild embarrassment that I did not see that solution.

Only concern is that it is a mild back compatibility break and if people are using a) the `pyplot` api b) `plt.ion` c) a non interactive backend this could break things if they do

```python
plt.plot(range(50))
plt.gca().get_xlim()   # this result changes under this behavior
```

**Comment by mdboom** [comment]
> Only concern is that it is a mild back compatibility break and if people are using a) the pyplot api b) plt.ion c) a non interactive backend this could break things if they do

That's a good point, and exactly the kind of thing I was hoping you'd help find.  I suppose what we ought to do is `if ion() ... do what we currently do in draw_idle ... else: pass`.

**Comment by WeatherGod** [comment]
Also, I would wonder if animations would break while in an AGG backend? I
can't remember if draw() or draw_idle() is used in animations.

**Comment by mdboom** [comment]
I've updated this so `draw_idle` is only a no-op if `interactive` is `False`.

@WeatherGod: I've confirmed that animation works with this change, both when saving to a file and "live" in TkAgg, Qt4Agg and NbAgg.

**Comment by mdboom** [comment]
@tacaswell: Not sure how to milestone this.  It's a pretty important performance improvement, but also has *some* potential for accidental consequences given how "core" it is.

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -2021,7 +2021,7 @@ def draw_idle(self, *args, **kwargs):
         """
         :meth:`draw` only if idle; defaults to draw but backends can overrride
```
Comment (jkseppan) [inline lib/matplotlib/backend_bases.py]: Perhaps mention the interactive-only behavior in the docstring?

File: lib/matplotlib/tests/test_text.py [hunk 2]
```diff
@@ -235,35 +239,40 @@ def test_axes_titles():
 
 @cleanup
 def test_set_position():
-    fig, ax = plt.subplots()
+    plt.ion()
```
Comment (jkseppan) [inline lib/matplotlib/tests/test_text.py]: Could we add a `plt.ioff()` call in the `cleanup` decorator?

**Comment by tacaswell** [comment]
I have been thinking about this a bit more recently and coming around to `draw_idle` should be a no-op for non-interactive backends.

If it is better to make this a no-op here and require that the sub-classes that want to provide it _do_ or to leave this as-is and make the backends that _do not_ want to provide it to over-ride with a no-op is not clear to me yet.

**Comment by jklymak** [comment]
This is very stale.  I can't tell if this is a good idea or not...

**Comment by dopplershift** [comment]
As far as I can tell, there's no opposition here, only unsure about any broader ramifications. This seems generally like a good idea to me...

**Comment by efiring** [comment]
The change in backend_bases looks good, and long overdue, though I haven't actually tried it and might be missing something.
I think @jkseppan's suggestion to add `plt.ioff()` to `cleanup` is good.  Then the last test would not need a special internal function, just a `plt.ion()`.

**Comment by tacaswell** [comment]
We have made other changes to the tests that I think actually eliminate most of the changes to the tests.

**Comment by tacaswell** [comment]
`(bleeding) ~/s/p/matplotlib dont-double-draw↑·18365↓·3|⚑18  `  Just a _bit_ out of date.

**Comment by tacaswell** [comment]
I think that this PR pre-dates the ability to push to contributor branches, I'll open a new one.

**Comment by tacaswell** [comment]
Closed in favor of #17923

**Pull request closed without merging**